[@types/hapi__joi] Update Joi to v17.1

stevendesu opened the pull request:
Per the template below, I still have some work to do to make sure things are kosher. But this should be fine.

Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [ ] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/hapijs/joi/blob/master/lib/index.js#L186
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

SimonSchick commented on types/hapi__joi/index.d.ts:
```diff
@@ -2054,6 +2054,11 @@ declare namespace Joi {
+        /**
+         * Checks whether or not the provided argument is an instance of ValidationError
+         */
+        isError(error: any): boolean;
```
Can you make `ValidationError` a proper class as it is exported as one and make `isError` return `error is ValidationError`?

stevendesu replied:
Updating `isError` to return `error is ValidationError` is easy and makes sense, but I'm not sure what all would go into making `ValidationError` "a proper class". Whether it's a class or an interface depends on implementation, not definition, right? And since the official `@hapi/joi` library just adds some properties to the base `Error` object, I don't know how defining it any differently than it is will make a difference.

Then again, I'm not a TypeScript guru, so maybe there's something I'm not understanding.

stevendesu replied:
@SimonSchick I looked into it and learned that there is a reason to declare `ValidationError` as a class: it enables you to say `new Joi.ValidationError()` within TypeScript without throwing an error. I also read the `@hapi/joi` source to verify that `Joi.ValidationError` ***was*** a class before continuing...

However after defining `Joi.ValidationError` as a class, a ton of other things broke in the definition. Apparently at the bottom of `hapi__joi/index.d.ts` there was a line re-declaring `Joi` in order to export `Joi.Root` instead of the whole `Joi` namespace:

```
declare namespace Joi {
    // ...
    interface Root {
        // ...
    }
    // ...
}

declare const Joi: Joi.Root;
export = Joi;
```

I don't know if this is a standard practice in TypeScript definitions (possibly to keep enums and other definitions hidden within the namespace and avoid polluting a global namespace), but the moment I changed `interface ValidationError extends Error {` to `class ValidationError extends Error {` this duplicate identify `Joi` started throwing errors (the duplicate identifier didn't throw an error when `ValidationError` was an interface)

Not sure how to proceed from here -- need someone with a better understanding of TypeScript to chime in.

SimonSchick replied:
Mhhh, ok that's a weird side effect, I will look into that.

SimonSchick replied:
Yea I can't get it to work either, the problem is that the `Joi` namespace only holds types UNTIL you add the class keyword when then makes it conflict with the declaration merge.

For now let's just go with `isError(error: any): error is ValidationError` and defer the class export to some other time.

stevendesu replied:
This morning I was derailed on another project, but before I make that change I wanted to read through the TypeScript docs and give this one more good attempt. See if I can understand how declaration files work, what's up with the `export = Joi` syntax (which doesn't look like valid JS to me), and see if I can understand what they were attempting to do with the `Joi.Root` interface and if I can refactor this to fix it

If I can't get it working in a couple of hours I'll give up and go with `isError(error: any): error is ValidationError`

stevendesu replied:
@SimonSchick so I'm about ready to give up on this. I read through the TypeScript docs and they weren't helpful -- 99% of the docs are for basic things like how to declare a function return type. Even among the basic things it was clear that the `@hapi/joi` definitions aren't following best practices, because the TypeScript docs said your definition files should have the same directory structure as your library, meaning we should have several files like `errors.d.ts`, `messages.d.ts`, `validator.d.ts`, etc.

I was going to just update the definition with `isError()` when I noticed something in the block comment at the top:

```
// Type definitions for @hapi/joi 16.0
```

I then checked the current version: https://hapi.dev/module/joi/

The latest is 17.1.1

Is there a proper process for updating this version number to avoid conflicts with anyone using DefinitelyTyped in the wild? If someone pulls in the `@types/hapi__joi` package and they're on v16 I don't want any changes I make that are v17-only to mess with them.

stevendesu commented:
In my second commit I've:

- Made `ValidationError` new-able, in lieu of being able to make it a proper class
- Added a `tslint:disable-next-line:no-misused-new` comment to bypass the error about new-able interfaces
- Updated the `isError` method to be a proper type guard
- Created a test for the `isError` method
- Updated the `declare const` values in the test to utilize the new-able interface (`const validErr = new Joi.ValidationError()`)
- Added myself to the list of contributors in the header

I'm uncertain of the proper way to proceed regarding the version number, however. The version number in the header is v16.0. The current version is v17.1. I don't know what implicit changes will occur if I update this (will it change the version of the NPM `@types/hapi__joi` module?) and I don't know if any functions or features were deprecated in Joi v17.1 that need to be **removed** from the definitions for them to still be valid.

For this reason, I would like input on how to proceed regarding the version number.

I also checked and the `isError` method was **not** available in Joi v16.0, so leaving the version unchanged is not a valid answer (as this will lead to invalid definition for v16)

SimonSchick commented on types/hapi__joi/index.d.ts:
```diff
@@ -608,6 +609,13 @@ declare namespace Joi {
+        // Discussion around this can be see in this pull request: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43970
+        // I really want a better solution, but this is better than nothing:
+        // tslint:disable-next-line:no-misused-new
+        new(): ValidationError;
```
Unfortunately this is incorrect, this would would allow you to use the `new` operator on an instance rather than on the class itself (which still doesn't exist).

I can see how you came up with this tho, `Root` incorrectly declares `ValidationError: ValidationError;`.
joi does not have a member `ValidationError` that is a validation error, it's the *class* `ValidationError`.

See:
![image](https://user-images.githubusercontent.com/3259089/79611553-b571fd00-80af-11ea-8443-ff1bf7b28fa3.png)

stevendesu replied:
I get the potential error of allowing the `new` operator to be used on an instance, but I don't think it's accurate that ***nothing*** should be visible when typing `Joi.ValidationError` into a TypeScript file, because there ***is*** a class with that name. As you just demonstrated in Node, `Joi.ValidationError` exists and, since it's a class, the following code is perfectly valid:

```
const Joi = require("@hapi/joi");
const myValidationError = new Joi.ValidationError();
```

After some testing I figured out I can emulate this behavior with the following change to Root:

```
ValidationError: new () => ValidationError;
```

It's not ideal (extremely ugly, actually) but it gets the types to behave the way you would expect.

Along with your other comments, I've been working on some additional local changes to get everything in sync with v17.1 (the version that added the `isError` method:

 - Replaced `wrapArrays` with the new `wrap` object containing `array` and `label` members
 - Updated the `default` method parameters to reflect the changes in v17
 - Added the `regex` method
 - Updated tests accordingly

I'll make a new commit shortly after running lint and test and everything else

SimonSchick replied:
Unfortunately this is a shortcoming of TS not being able to map out all the existing (and sometimes crazy) export patterns people in JS, even the node typings have exceptions like this 😛

stevendesu commented:
Upon realizing that the `isError` method was added in v17.1 I've gone through the changelog and brought the other definitions up to speed with v17.1 and amended the tests accordingly, and bumped the version number.

stevendesu commented:
I want to update some additional types and realized that operating on `master` on my own repository was problematic. I've moved my changes to their own branch but can't find a way to update this PR to point to the other branch.

I'll avoid updating my `master` on GitHub for now, but does anyone know how to change which branch a PR is targeting?

SimonSchick commented:
Unfortunately this doesn't seem to work, you can only change the target branch.
You can keep the master in this state and still work in other branch however:

```
git fetch upstream // assuming you called it upstream :)
git checkout -b newbranch // checkout new branch
git reset --hard upstream/master // resets your branch to be at the same commit as DT master
```

But yea, in the future you want to avoid committing to your master :)

weswigham commented:
@stevendesu where do you wanna take this? Is this up to date?

stevendesu commented:
@weswigham In d5b91c2528de5bedb6afd0693ab3e212c9eb23e8 I went through the Joi release notes and added / changed anything per their update guide -- so it should be up to date if their release notes are exhaustive.
